packages/react-utilities/src/compose/types.ts

```diff
@@ -16,17 +16,64 @@ export type ShorthandProps<Props extends DefaultObjectShorthandProps> =
   | undefined // default render (or null render if no default provided)
   | Props;
 
-export type DefaultObjectShorthandProps = ObjectShorthandProps<{}, unknown, keyof JSX.IntrinsicElements>;
-
-export type ObjectShorthandProps<
-  Props extends { children?: React.ReactNode } = {},
-  Ref = unknown,
-  As extends keyof JSX.IntrinsicElements = never
-> = Props &
-  React.RefAttributes<Ref> & {
-    as?: As;
-    children?: Props['children'] | ShorthandRenderFunction<Props>;
-  };
+/**
+ * Matches any shorthand props type.
+ *
+ * This should ONLY be used in type templates as in `extends DefaultObjectShorthandProps`;
+ * it shouldn't be used as the type of a slot.
+ */
+export type DefaultObjectShorthandProps = ObjectShorthandProps<{
+  children?: React.ReactNode;
+  as?: keyof JSX.IntrinsicElements;
+}>;
+
+/**
+ * Defines the slot props for a slot that supports a Component type.
+ *
+ * For intrinsic elements like 'div', use {@link ElementShorthandProps} instead.
+ */
+export type ObjectShorthandProps<Props extends { children?: React.ReactNode } = {}> = Props & {
+  children?: Props['children'] | ShorthandRenderFunction<Props>;
+};
+
+/**
+ * Define the slot arguments for a slot that supports one or more intrinsic element types, such as 'div'.
+ * For slots that support custom components, use {@link ObjectShorthandProps} instead.
+ *
+ * The first param is the slot's default type if no `as` prop is specified.
+ * The second param is an optional union of alternative types that can be specified for the `as` prop.
+ *
+ * ```
+ * ElementShorthandProps<'div'> // Slot is always a <div>
+ * ElementShorthandProps<'a', 'button'> // Defaults to <a>, but allows { as: 'button' } to be a <button>
+ * ElementShorthandProps<'label', 'span' | 'div'>; // Defaults to <label>, but allows { as: 'span' } or { as: 'div' }
```

Comment:
This looks interesting, will the props adjust based on the value of the `as` prop ?

Reply:
From my understanding yes, that is what is supposed to happen.

Reply:
Yes. This is a [discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) with `as` being the "discriminant" property. When you write `{ as: 'button', ... }`, Typescript knows that the rest of the props can only come from button's props.